phoenix-core/src/it/java/org/apache/phoenix/end2end/IndexUpgradeToolIT.java

```diff
@@ -0,0 +1,506 @@
+        conn.createStatement().execute("CREATE INDEX INDEX3 ON TEST.MOCK3 (age, name)");
+    }
+
+    @Parameters(name ="IndexUpgradeToolIT_mutable={0},upgrade={1}")
+    public static Collection<Object[]> data() {
+        return Arrays.asList(new Object[][] {
+                {false, true},
+                {true, true},
+                {false, false},
+                {true, false}
+        });
+    }
+
+    public IndexUpgradeToolIT (boolean mutable, boolean upgrade) {
+        this.mutable = mutable;
+        this.upgrade = upgrade;
+    }
+
+    @Test
+    public void testToolWithOneTableOneIndex() throws SQLException, IOException {
+        String schemaName = generateUniqueName();
+        String dataTableName = generateUniqueName();
+        String dataTableFullName = SchemaUtil.getTableName(schemaName, dataTableName);
+        String indexTableName = generateUniqueName();
+        try (Connection conn = DriverManager.getConnection(getUrl(), new Properties())) {
+            conn.setAutoCommit(true);
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class).getQueryServices();
+            Admin admin = queryServices.getAdmin();
+            conn.createStatement().execute("CREATE TABLE " + dataTableFullName + " (id bigint NOT NULL PRIMARY KEY, name varchar)" + tableDDLOptions);
+            conn.createStatement().execute("CREATE INDEX " + indexTableName + " ON " + dataTableFullName + " (name)");
+
+            ArrayList<String> args = new ArrayList<>();
+            args.add("-o");
+            if (upgrade) {
+                args.add("upgrade");
+                if (mutable) {
+                    Assert.assertTrue(admin.getDescriptor(TableName.valueOf(dataTableFullName)).hasCoprocessor(Indexer.class.getName()));
+                    Assert.assertFalse(admin.getDescriptor(TableName.valueOf(dataTableFullName)).hasCoprocessor(IndexRegionObserver.class.getName()));
+                }
+                Assert.assertFalse(admin.getDescriptor(TableName.valueOf(schemaName + "." + indexTableName)).hasCoprocessor(GlobalIndexChecker.class.getName()));
+            } else {
+                args.add("rollback");
+                if (mutable) {
+                    Assert.assertTrue(admin.getDescriptor(TableName.valueOf(dataTableFullName)).hasCoprocessor(IndexRegionObserver.class.getName()));
+                    Assert.assertFalse(admin.getDescriptor(TableName.valueOf(dataTableFullName)).hasCoprocessor(Indexer.class.getName()));
+                }
+                Assert.assertTrue(admin.getDescriptor(TableName.valueOf(schemaName + "." + indexTableName)).hasCoprocessor(GlobalIndexChecker.class.getName()));
+            }
+            args.add("-tb");
+            args.add(dataTableFullName + ":" + indexTableName);
+            args.add("-c");
+            args.add(conn.unwrap(PhoenixConnection.class).getURL());
+            IndexUpgradeTool.main(args.toArray(new String[args.size()]));
+            if (upgrade) {
```

Comment:
This validation block from 156-168 looks like a special-case of the same logic in next test fn. Refactor to reuse the same code?

Reply:
And looks like all tests can share the code if we passed an array/list of indexes to verify. Will make you test file smaller.